umap/static/umap/js/modules/data/layer.js

```diff
@@ -36,7 +36,7 @@ const LAYER_MAP = LAYER_TYPES.reduce((acc, klass) => {
 }, {})
 
 export class DataLayer {
-  constructor(map, data) {
+  constructor(map, data, sync, future_uuid) {
```

Comment:
I don't get why `sync` is needed here.

Reply:
Also, at this time, I did not understand the need of `future_uuid`, can you elaborate a bit on why we can't just have an id/uuid ?

Reply:
>  I did not understand the need of future_uuid / why we can't just have an id/uuid ?

We need a way to distinguish between three different scenarii here:

1. Creation happened locally, but is not yet synced on the server ;
2. Creation is coming from a remote peer ;
3. Creation is coming from an already saved map, with existing layers.

`future_uuid` is meant to solve the first two, when a layer is created locally on the map, but the server isn't yet aware of it. In this case, we need to decide on the uuid locally, but we need a way to know that the server doesn't know yet about it, so we can ask it to save. That's the purpose of this field.

In an earlier iteration, I tried to track the state of the server by using the `version` info provided by the server, but I believe this design is better because it's more explicit that the uuid isn't yet saved.

>  don't get why sync is needed here.

Ah :-) ` sync` is a flag to tell you if you need to replicate the creation on other peers via the sync protocol. It is not used at the moment, but it is passed by `map.createDataLayer()`. I can remove it!

Reply:
>  I can remove it!

Thanks! I'd prefer the `sync` attribute to stay outside data classes as much as possible

> 2.  Creation is coming from a remote peer

Isn't actually what the `sync` attribute is for ? Or otherwise I don't get why we need to distinguish this case from the others (what do we need to do/prevent for this case ?).

I'm in favor of the goal of this PR, which is to move the uuid creation from the backend to the frontend. Doing so, I'm expecting that the front client that creates a new datalayer is the source of truth, and both other clients and server should follow.

And even for the other cases, it's not clear to me why we need to distinguish them. For me, always doing a PUT to the server would be good (as we are the source of truth when we save now), and then when creating a datalayer with data coming from the server, I'd expect that the server only act as a storage, so as soon as the datalayer is created on the frontend, it behave like any datalayer.

I feel like I'm missing something obvious, but I can't find it ;)

Reply:
I think what I did in this PR is one way to do it. It's not perfect, but it works without adding complexity to the overall design.

> Isn't actually what the `sync` attribute is for ? Or otherwise I don't get why we need to distinguish this case from the others (what do we need to do/prevent for this case ?).

Sort of, but `sync` is missing part of the info I need: the fact that it's not saved yet on the server. This is not provided by the `sync` property, as it's just saying that it's coming from another peer. It doesn't say if that peer has saved the data yet (or not).

I think that `_future_uuid` is an elegant way to solve this: it stores the uuid so it's possible to pass it to the server on save, and and the same time it's telling us that this datalayer hasn't been saved yet on the server (info that was provided before by looking at `this.umap_id !== undefined`.

> And even for the other cases, it's not clear to me why we need to distinguish them. For me, always doing a PUT to the server would be good (as we are the source of truth when we save now), and then when creating a datalayer with data coming from the server, I'd expect that the server only act as a storage, so as soon as the datalayer is created on the frontend, it behave like any datalayer.

I agree that it would be the ultimate goal to merge the two `/create/` and `/update` endpoints.

I will have another look at this tomorrow, to see if it's possible to merge the two `DataLayerCreate` and `DataLayerUpdate` classes together. The reason I didn't do it yet, is because it feels to be a bigger step.

With that being said, having these two endpoints merged in one might not remove the requirement of the `future_uuid` field, as we will still need to communicate about the datalayer (with other peers) before it's saved to the server, **and** convey the fact that it's not saved yet.

> it's not clear to me why we need to distinguish them.

Why do we need to know that it's not saved yet? For the same reasons we check if `umap_id` is defined or not: because if it's not saved we don't make a request to the server asking for it. Instead, we create it.

> I'd expect that the server only act as a storage, so as soon as the datalayer is created on the frontend, it behave like any datalayer.

I agree, and I believe that's what's happening right now in this PR 🙌

Reply:
Sorry, I still don't get the `_future_uuid` pattern :(

But I'm certainly missing the problem we are trying to solve.

For what I understand, I'd:

- make any peer always PUT (when dirty, of course)
- on reset/delete, always send a DELETE to the server, and don't care if it's a 404 (or rely on `reference_version` if we want to be explicit, but I'm not sure it's worth it)

I'm also tempted to not sync to other peers the datalayer unless it's persisted (if client A creates a datalayer without saving it, then sync, then client B adds data in this datalayer without saving, then client A cancel the datalayer creation, what happen to client B data ?).

Reply:
> I'm also tempted to not sync to other peers the datalayer unless it's persisted (if client A creates a datalayer without saving it, then sync, then client B adds data in this datalayer without saving, then client A cancel the datalayer creation, what happen to client B data ?).

The delete of the datalayer will also be synced, so we're good here :+1:

Reply:
> But I'm certainly missing the problem we are trying to solve.

The problem is: we want the ability to sync layers creation and property update before server-save. 

It's not at odds to what you're proposing here:

> For what I understand, I'd:
> 
> - make any peer always PUT (when dirty, of course)
> - on reset/delete, always send a DELETE to the server, and don't care if it's a 404 (or rely on reference_version if we want to be explicit, but I'm not sure it's worth it)

As I said yesterday I want to have another look at how to merge the two views together, to always do a PUT.

The only bit that stays in the way is the `_future_uuid` field, and merging the two views won't make the `_future_uuid` **concept** go away.

The current code looks at `dl.umap_id == undefined` in a few places and uses this information to act differently pre-save. I believe these cases to be here for a reason, and `_future_uuid` is just a way to not get in the way of it. We could do this with another field like `not_saved_on_server` or similar, but that will stay conceptually the same.